scripts/dependencies/testing.py

```diff
@@ -0,0 +1,69 @@
+#!/usr/bin/env python
+"""
+List any dependencies on test modules in edx-platform from non-test modules.
+Generally, there shouldn't be any; such a dependency could result in
+ImportErrors in production where testing packages aren't installed.
+
+This script counts on scripts/dependencies/enumerate.sh having already
+been run in order to generate a dependency data file to work from.
+"""
+from __future__ import absolute_import, print_function
+
+import re
+
+# Enumerate all the Python modules that should only be imported during test runs
+pattern_fragments = [
+    # Test modules within edx-platform
+    r'/tests?\.py',                            # test.py, tests.py
+    r'/tests?_[^/]*\.py',                      # test_*.py, tests_*.py
+    r'/[^/]*_tests\.py',                       # *_tests.py
+    r'/tests?/',                               # */test/*, */tests/*
+    r'[cl]ms/.*/features/',                    # cms/*/features/*, lms/*/features/*
+    r'/testing\.py',                           # testing.py
+    r'/testutils\.py',                         # testutils.py
+    r'/tests$',                                # tests/__init__.py
+    r'conftest\.py',                           # conftest.py
+    r'/envs/acceptance\.py',                   # cms/envs/acceptance.py, lms/envs/acceptance.py
+    r'/envs/acceptance_docker\.py',            # cms/envs/acceptance.py, lms/envs/acceptance.py
+    r'/factories\.py',                         # factories.py
+    r'^terrain',                               # terrain/*
+    r'/setup_models_to_send_test_emails\.py',  # setup_models_to_send_test_emails management command
+
+    # Testing-only package dependencies
+    r'^bs4',                                   # beautifulsoup4
```

Comment:
I considered trying that, but I'm concerned that:

1. It could miss dependencies on packages which use different names for the directory and the PyPI listing.  The only tricky case in this set would be bok-choy vs. bok_choy, which could be dealt with easily enough.
2. It would add a bunch of extra logic to the regex which doesn't actually correspond to any real packages (like pytest plugins), slowing down the script.
3. A testing dependency could theoretically be a subset of one of our own top-level directory names, although that seems unlikely.

Do you think it's worth auto-generating this and accepting these risks?

Reply:
I think in a trade-off between those risks and having to manually keep them in sync I'd lean toward the manual solution. I agree that the use cases are potentially complicated enough that it's better to be explicit.